proposals/0155-normalize-enum-case-representation.md

```diff
@@ -181,58 +181,88 @@ enum Tree {
+case .buffalo(_): // ok
+case .buffalo(_, _): // ok
 }
 ```
 
-Compared to patterns in Swift 3, matching against enum cases will follow
-stricter rules. This is a consequence of no longer relying on tuple patterns.
+To maintain source compatibility, this behavior cannot be changed. However, it
+poses an ambiguity when combined with the added functionality above, where
+multiple cases share the same base name, but have different argument labels
+or number of arguments:
 
-When an associated value has a label, the sub-pattern must include the label
-exactly as declared. There are two variants that should look familiar to Swift
-3 users. Variant 1 allows user to bind the associated value to arbitrary name in
-the pattern by requiring the label:
+```
+enum Buffalo {
+    case buffalo(String)
+    case buffalo(bill: String)
+    case buffalo(buffalo: String)
+    case buffalo(buffalo: String, buffalo: String)
+}
 
-```swift
-case .variable(name: let x) // okay
-case .variable(x: let x) // compile error; there's no label `x`
-case .lambda(parameters: let params, body: let body) // Okay
-case .lambda(params: let params, body: let body) // error: 1st label mismatches
+switch buffalo {
+case .buffalo(_): // Ambiguous under the original rules, could refer to any of the above
+}
 ```
 
-User may choose not to use binding names that differ from labels. In this
-variant, the corresponding value will bind to the label, resulting in this
-shorter form:
+To resolve the ambiguity, single-element patterns shall favor matching single-
+element associated values over tuples. Explicit destructuring must be used to
+favor a multiple-element associated value:
 
-```swift
-case .variable(let name) // okay, because the name is the same as the label
-case .lambda(let parameters, let body) // this is okay too, same reason.
-case .variable(let x) // compiler error. label must appear one way or another.
-case .lambda(let params, let body) // compiler error, same reason as above.
 ```
+enum Buffalo {
+    case buffalo(String)
+    case buffalo(buffalo: String, buffalo: String)
+}
 
-Only one of these variants may appear in a single pattern. Swift compiler will
-raise a compile error for mixed usage.
+switch buffalo {
+case .buffalo(_): // Favors `buffalo(_:)`
+case .buffalo(_, _): // Favors `buffalo(buffalo:buffalo:)`
+}
+```
+
+
+If multiple enum cases agree in arity but have different argument labels, then
+an unlabeled pattern favors matching an unlabeled associated value:
 
-```swift
-case .lambda(parameters: let params, let body) // error, can not mix the two.
 ```
+enum Buffalo {
+    case buffalo(String, String)
+    case buffalo(buffalo: String, buffalo: String)
+}
 
-Some patterns will no longer match enum cases. For example, all associated
-values can bind as a tuple in Swift 3, this will no longer work after this
-proposal:
+switch buffalo {
+case .buffalo(_, _): // Favors `buffalo(_:_:)`
+case .buffalo(_, _): // Favors `buffalo(buffalo:buffalo:)`
+}
```

Comment:
Is this a typo, or is it suggesting that the type checker will be super-extra-smart in preferring `buffalo(buffalo:buffalo:)` in the second `case` because it would have "realized" that the otherwise default preferred match (`buffalo(_:_:)`) is already covered by the first `case`?

Reply:
It's a typo. Thanks!

Reply:
Fixed in https://github.com/apple/swift-evolution/commit/23769674c62b9aa0491fe91b89cd2d3c936ab599